### PR TITLE
Set proper alignment on constants

### LINKED ISSUE #28912: LLVM lint: Memory reference address is misaligned
This is a subissue of #7463.

```rust
pub enum H<A, B> {
    L(A),
    R(B),
}
pub fn test() -> H<i32, i64> {
    H::L(0)
}
```

Compile with `rustc -C passes=lint` and you get:

```
Undefined behavior: Memory reference address is misaligned
  call void @llvm.memcpy.p0i8.p0i8.i64(i8* %2, i8* getelementptr inbounds ({ i8, i32, [8 x i8] }, { i8, i32, [8 x i8] }* @const3020, i32 0, i32 0), i64 16, i32 8, i1 false)
```

Manually setting alignment of const3020 to 8 in LLVM IR fixes the lint.

### PR DESCRIPTION
**Pull request by dotdash**
For enum variants, the default alignment for a specific variant might be
lower than the alignment of the enum type itself. In such cases we, for
example, generate memcpy calls with an alignment that's higher than the
alignment of the constant we copy from.

To avoid that, we need to explicitly set the required alignment on
constants.

Fixes #28912.

### CODE REVIEW

File: src/librustc_trans/trans/consts.rs [hunk 1]
```diff
@@ -254,11 +254,11 @@ pub fn get_const_expr_as_global<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>,
         Some(&val) => return val,
         None => {}
     }
+    let ty = monomorphize::apply_param_substs(ccx.tcx(), param_substs,
+    &ccx.tcx().expr_ty(expr));
```
Comment (eddyb) [inline src/librustc_trans/trans/consts.rs]: Alignment mishap?

**Comment by eddyb** [comment]
r=me with the formatting nit fixed.

File: src/librustc_trans/trans/consts.rs [hunk 2]
```diff
@@ -275,6 +275,10 @@ pub fn get_const_expr_as_global<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>,
     };
 
     let lvalue = addr_of(ccx, val, "const");
+    unsafe {
+        llvm::LLVMSetAlignment(lvalue, type_of::align_of(ccx, ty));
```
Comment (eefriedman) [inline src/librustc_trans/trans/consts.rs]: I think this belongs inside addr_of_mut, rather than here.

**Comment by dotdash** [comment]
@bors r-

@eefriedman Good point.

**Comment by dotdash** [comment]
Update as per the suggestion of @eefriedman. I also had to fix the string that identifies the gdb script section, which is used as a C string but had no terminating 0 byte. The alignment changes caused that bug to trigger an error in a codegen test.

File: src/librustc_trans/trans/debuginfo/gdb.rs [hunk 3]
```diff
@@ -54,7 +54,8 @@ pub fn insert_reference_to_gdb_debug_scripts_section_global(ccx: &CrateContext)
 /// section.
 pub fn get_or_insert_gdb_debug_scripts_section_global(ccx: &CrateContext)
                                                   -> llvm::ValueRef {
-    let section_var_name = "__rustc_debug_gdb_scripts_section__";
+    let c_section_var_name = "__rustc_debug_gdb_scripts_section__\0";
+    let section_var_name = &c_section_var_name[..c_section_var_name.len()-1];
```
Comment (eefriedman) [inline src/librustc_trans/trans/debuginfo/gdb.rs]: For clarity, you might want to use `c_section_var_name.as_ptr() as *const _` below.

File: src/librustc_trans/trans/consts.rs [hunk 4]
```diff
@@ -130,13 +132,14 @@ fn addr_of_mut(ccx: &CrateContext,
 
 pub fn addr_of(ccx: &CrateContext,
                cv: ValueRef,
+               align: machine::llalign,
                kind: &str)
                -> ValueRef {
     match ccx.const_globals().borrow().get(&cv) {
         Some(&gv) => return gv,
```
Comment (eefriedman) [inline src/librustc_trans/trans/consts.rs]: I think you need to do some sort of alignment check here?

File: src/librustc_trans/trans/debuginfo/gdb.rs (hunk 3 above)
Comment (dotdash) [inline src/librustc_trans/trans/debuginfo/gdb.rs]: For correctness even. Not sure where I lost that change (and even more confusing that it passed the test without it). Thanks!

File: src/librustc_trans/trans/consts.rs (hunk 4 above)
Comment (dotdash) [inline src/librustc_trans/trans/consts.rs]: Not sure what you mean.

Comment (eefriedman) [inline src/librustc_trans/trans/consts.rs]: A global could be found in the `const_globals` map which has the correct value, but not the correct alignment.

Comment (dotdash) [inline src/librustc_trans/trans/consts.rs]: Ah, so just an assert? I thought there was something missing that would cause misalignment or something.

Comment (eefriedman) [inline src/librustc_trans/trans/consts.rs]: No, not just an assert: you can legitimately end up with a value with the wrong alignment.  Consider:

```rust
#![crate_type="lib"]
#[repr(i16)]
pub enum E<A, B> {
    A(A),
    B(B),
}

pub fn f() -> E<i16, [i16; 3]> {
    E::A(0)
}

pub fn g() -> E<i16, i32> {
     E::A(0)
}
```

Comment (dotdash) [inline src/librustc_trans/trans/consts.rs]: Oh, right, it looks them up by value not by name. Sorry for being dense here. I'll fix that to upgrade the alignment as required and add a test for it.

**Comment by dotdash** [comment]
Updated to handle upgrading the alignment of constants that are shared between types with different alignment requirements.

Thanks a lot @eefriedman!

**Pull request merged**